httpx/dispatch/proxy_http.py

```diff
@@ -0,0 +1,236 @@
+import enum
+
+import h11
+
+from ..concurrency.base import ConcurrencyBackend
+from ..config import (
+    DEFAULT_POOL_LIMITS,
+    DEFAULT_TIMEOUT_CONFIG,
+    CertTypes,
+    PoolLimits,
+    SSLConfig,
+    TimeoutTypes,
+    VerifyTypes,
+)
+from ..exceptions import ProxyError
+from ..middleware.basic_auth import build_basic_auth_header
+from ..models import (
+    URL,
+    AsyncRequest,
+    AsyncResponse,
+    Headers,
+    HeaderTypes,
+    Origin,
+    URLTypes,
+)
+from ..utils import get_logger
+from .connection import HTTPConnection
+from .connection_pool import ConnectionPool
+from .http2 import HTTP2Connection
+from .http11 import HTTP11Connection
+
+logger = get_logger(__name__)
+
+
+class HTTPProxyMode(enum.Enum):
+    DEFAULT = "DEFAULT"
+    FORWARD_ONLY = "FORWARD_ONLY"
+    TUNNEL_ONLY = "TUNNEL_ONLY"
+
+
+class HTTPProxy(ConnectionPool):
+    """A proxy that sends requests to the recipient server
+    on behalf of the connecting client.
+    """
+
+    def __init__(
+        self,
+        proxy_url: URLTypes,
+        *,
+        proxy_headers: HeaderTypes = None,
+        proxy_mode: HTTPProxyMode = HTTPProxyMode.DEFAULT,
+        verify: VerifyTypes = True,
+        cert: CertTypes = None,
+        timeout: TimeoutTypes = DEFAULT_TIMEOUT_CONFIG,
+        pool_limits: PoolLimits = DEFAULT_POOL_LIMITS,
+        backend: ConcurrencyBackend = None,
+    ):
+
+        super(HTTPProxy, self).__init__(
+            verify=verify,
+            cert=cert,
+            timeout=timeout,
+            pool_limits=pool_limits,
+            backend=backend,
+        )
+
+        self.proxy_url = URL(proxy_url)
+        self.proxy_mode = proxy_mode
+        self.proxy_headers = Headers(proxy_headers)
+
+        url = self.proxy_url
+        if url.username or url.password:
+            self.proxy_headers.setdefault(
+                "Proxy-Authorization",
+                build_basic_auth_header(url.username, url.password),
+            )
+            self.proxy_url = url.copy_with(authority=url.authority.rpartition("@")[2])
```

Comment:
Is this related to https://github.com/encode/httpx/issues/328? Should we add a comment?

Reply:
No, this is how to split userinfo from authority. Rfc3986 package doesn't provide an easy way to work with subauthority components. :(

Reply:
I'm a bit confused.

*If* this could just read `url.copy_with(authority=url.<some_named_property>)` then what is `<some_named_property>` here? Are we missing a property that we ought to add on our `URL` model?

Reply:
It would be the authority minus the auth section. I don't think we're missing any component, just the underlying rfc3986 library doesn't provide the tools to remove the userinfo section cleanly.